x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/PolicyStepsRegistry.java

```diff
@@ -145,103 +126,83 @@ public LifecyclePolicyMetadata read(StreamInput in, String key) {
+        }
+        LifecyclePolicySecurityClient policyClient = new LifecyclePolicySecurityClient(client,
+            ClientHelper.INDEX_LIFECYCLE_ORIGIN, lifecyclePolicyMap.get(policy).getHeaders());
+        final List<Step> steps = policyToExecute.toSteps(policyClient);
+        // Build a list of steps that correspond with the phase the index is currently in
+        final List<Step> phaseSteps;
+        if (steps == null) {
+            phaseSteps = new ArrayList<>();
+        } else {
+            phaseSteps = steps.stream()
+                .filter(e -> e.getKey().getPhase().equals(currentPhase))
+                .collect(Collectors.toList());
         }
+        logger.trace("parsed steps for policy [{}] in phase [{}], definition: [{}], steps: [{}]",
+            policy, currentPhase, phaseDef, phaseSteps);
+        return phaseSteps;
     }
 
-    /**
-     * returns the {@link Step} that matches the index name and
-     * stepkey specified. This is used by {@link ClusterState}
-     * readers that know the current policy and step by name
-     * as String values in the cluster state.
-     * @param index the index to get the step for
-     * @param stepKey the key to the requested {@link Step}
-     * @return the step for the given stepkey or null if the step was not found
-     */
     @Nullable
-    public Step getStep(final Index index, final Step.StepKey stepKey) {
+    public Step getStep(final IndexMetaData indexMetaData, final Step.StepKey stepKey) {
         if (ErrorStep.NAME.equals(stepKey.getName())) {
             return new ErrorStep(new Step.StepKey(stepKey.getPhase(), stepKey.getAction(), ErrorStep.NAME));
         }
 
-        if (indexPhaseSteps.get(index) == null) {
-            return null;
+        final String phase = stepKey.getPhase();
+        final String policyName = indexMetaData.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+        final Index index = indexMetaData.getIndex();
+
+        if (policyName == null) {
+            throw new IllegalArgumentException("failed to retrieve step " + stepKey + " as index [" + index.getName() + "] has no policy");
         }
 
-        if (logger.isTraceEnabled()) {
-            logger.trace("[{}]: retrieving step [{}], found: [{}]\nall steps for this phase: [{}]", index, stepKey,
-                indexPhaseSteps.get(index).stream().filter(step -> step.getKey().equals(stepKey)).findFirst().orElse(null),
-                indexPhaseSteps.get(index));
-        } else if (logger.isDebugEnabled()) {
-            logger.debug("[{}]: retrieving step [{}], found: [{}]", index, stepKey,
-                indexPhaseSteps.get(index).stream().filter(step -> step.getKey().equals(stepKey)).findFirst().orElse(null));
+
+        // parse phase steps from the phase definition in the index settings
+        final String phaseJson = indexMetaData.getSettings().get(LifecycleSettings.LIFECYCLE_PHASE_DEFINITION,
```

Comment:
This will also need to be updated for #33783, depending on which gets merged first.